### DISCUSSION TITLE
Private Procedures, Functions, and Macros

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#8220] [Ideas]
Whan you have a tiddler contains procedures/functions/macros and tagged with $:/tags/Global or $:/tags/Macros every routine is global. Sometimes you want sub procedures/functions/macros which are used by other procedures/functions/macros. In this case you like to have private procedures/functions/macros.

One way is: to use NOT global procedures/functions/macros are defined in separate tiddlers and then import them inside procedures/functions/macros.

My idea is: can we use double underscore or similar format in the name of such procedures/functions/macros and make them private?

**Comment by saqimtiaz** [comment]
Place the private procedures inside the global one.

**Reply by kookma** [reply]
Hi @saqimtiaz 
That is quite true. I use nested procedure/function as much as possible and it works great.
There is one issue however, when you use a sub-procedure in more than on global procedure/function.

**Reply by saqimtiaz** [reply]
> There is one issue however, when you use a sub-procedure in more than on global procedure/function.

For such scenarios I place the shared variables in a separate tiddler and import them inside the global procedure

**Comment by pmario** [comment]
The main problem I do have with private code is that it is close to impossible to modify be 3rd party programmers. So everything basically is hardcoded and changing something will need to overwrite the whole global code. 

Debugging private code also is a pain, since it cannot be easily reached, except if it is \imported. -- But imo managing imported content also is painful within TW editors. There is no fast way to open `\import`ed code tiddlers. 

So for my code I try to avoid it. Having well defined procedure and function names helps to avoid naming clashes and it forces me to really think about "self documenting" names. Which imo is an advantage for maintenance.

Just some thoughts -m

**Reply by saqimtiaz** [reply]
> The main problem I do have with private code is that it is close to impossible to modify be 3rd party programmers.

I find that desirable in my code and a boon for managing support requests, along with providing well designed mechanisms for users to override and tweak behaviour in appropriate places.
